Send multiple messages to WebSocket sources

benmanns opened the pull request:
This updates the WebSocket connector to add a `subscription_messages` (plural) config to `websocket_config`, which allows sending multiple subscription messages to a WebSocket source endpoint.

I have made some design decisions here that I'm open to modification on:
* `subscription_messages` is a new config key separate from `subscription_message`. If `subscription_message` is used, a deprecation warning appears suggesting using `subscription_messages`.
  * Alternative: change the type to an enum or similar as necessary to support a single string or array of strings and accept both under the `subscription_message` key. Not sure if possible with the YAML parser—I didn't try.
  * Alternative: accept both, with `subscription_message` appearing at the beginning or end of the message list.

I find this necessary for third-party WebSocket sources such as the [Polygon.io Options WebSocket endpoint](https://polygon.io/docs/options/ws_getting-started), which requires at least two messages, one for authentication and one per data channel.

For reference, my connection config for the Polygon.io options data looks like:

```yaml
meta:
  version: latest
  name: polygon-options-websocket-connector
  type: websocket-source
  topic: {TOPIC}
http:
  endpoint: wss://socket.polygon.io/options
  websocket_config:
    subscription_messages:
      - '{"action":"auth","params":"{POLYGON_API_KEY}"}'
      - '{"action":"subscribe","params":"T.*,A.*,AM.*"}'
```

digikata commented on crates/http-source/src/websocket_source.rs:
```diff
@@ -28,7 +28,7 @@ pub(crate) struct WebSocketSource {
 struct WSRequest {
     url: Url,
-    subscription_message: Option<String>,
+    subscription_messages: Option<Vec<String>>,
```
Instead of carrying a Option and a Vec, it's probably cleaner to make this a plain Vec.  This makes the code later on just iterate over the vec. If the vec is empty, then the loop doesn't send anything.

benmanns replied:
Updated to `Vec<String>` that defaults to `vec![]` if not provided.

digikata commented on crates/http-source/src/websocket_source.rs:
```diff
@@ -128,11 +130,23 @@ async fn websocket_writer_and_stream<'a>(
 impl WebSocketSource {
     pub(crate) fn new(config: &HttpConfig) -> Result<Self> {
         let ws_config = config.websocket_config.as_ref();
+        let subscription_messages = if let Some(ws_config) = ws_config {
+            if let Some(messages) = ws_config.subscription_messages.as_ref() {
+                Some(messages.clone())
+            } else if let Some(message) = ws_config.subscription_message.as_ref() {
+                warn!("`subscription_message` is deprecated, please use `subscription_messages` instead");
+                Some(vec![message.to_owned()])
+            } else {
+                None
+            }
+        } else {
+            None
+        };
```
I think lets accept both for now & keep the deprecation message.  I would suggest always start with an empty vec and push or append any supplied values for `subscription_message` or `subscription_messages`, with `subscription_message` at the beginning if provided.

This pr should probably also bump the connector version to 0.4.0 and refresh the README.md with that version too.  Between this and the other PR it's a nice upgrade!

benmanns replied:
Updated to the `Vec<String>`, defaulting to `vec![]`, if `subscription_message` is provided then `vec![subscription_message]`, and if both then `vec![subscription_message] + subscription_messages`.

digikata reviewed (changes requested):
I would suggest that both are accepted,

benmanns commented:
Thanks for the feedback! I'll address your comments and the merge conflicts between the two branches shortly.

benmanns commented:
@digikata thanks for the review! I addressed the merge conflicts from #244, your comments, updated the connector documentation in the README with the new config option, and bumped to v0.4.0.

benmanns commented on README.md:
```diff
@@ -15,18 +15,19 @@ See [docs](https://www.fluvio.io/connectors/inbound/http/) here.
-| output_parts | body                       | String  | `body` = body only, `full` = all status, header and body parts                             |
-| stream       | false                      | bool    | Flag to indicate HTTP streaming mode                                                       |
-| delimiter    | '\n'                       | String  | Delimiter to separate records when producing from an HTTP streaming endpoint               |
+| Option           | default                    | type            | description                                                                                |
```
There's a lot of lines changed here to keep the table aligned, but the only modified entry is the websocket_config line.

digikata commented:
Looks good @benmanns, thanks for the changes!

ajhunyady commented:
It would be great to add a message using new newly added field `subscription_messages`.
